jumanji/environments/routing/op/env.py

```diff
@@ -0,0 +1,313 @@
+            a chosen node is too far to reach the depot within the time budget.
+            
+    
+    - state: 'State'
+        - coordinates: jax array (float) of shape (num_nodes + 1, 2)
+            the coordinates of each node.
+        - position: jax array (int32) of shape ()
+            the index corresponding to the last visited node. 
+        - visited_mask: jax array (bool) of shape (num_nodes + 1, )
+            binary mask (False/True <--> not visited/visited).    
+        - trajectory: jax array (int32) of shape (num_nodes, )
+            array of node indicies defining the route (set to DEPOT_IDX if not filled yet).
+        - num_visited: int32
+                number of total nodes visited
+        - prizes: jax array (float) of shape (num_nodes + 1, ) 
+            the associated prizez of each node and the depot note (0.0 for the depot)
+        - length: jax array (float) of shape (num_nodes + 1, )
+            the length between each node and the depot (0.0 for the depot)  
+        -remaining_max_length: jax array (float) of shape ()
+            the remaining length budget  
+            
+    
+    [1] Wouter Kool, Herke van Hoof, and Max Welling. (2019). "Attention, learn to solve routing problems!" 
+
+    """
+    
+    
+    def __init__(
+        self,
+        generator: Optional[generator.Generator] = None,
+        reward_fn: Optional[reward.RewardFn] = None,
+        viewer: Optional[Viewer[types_.State]] = None,
+    ):
+        """Instantiates an OP environment.
+        
+        
+        Args:
+            generator: 'Generator' whose '__call__' method instantiates an environment instance.
+                The default option is 'UniformGenerator' which randomly generates OP instances 
+                with 20 nodes and prizes sampled from a uniform distribution .
+            reward_fn: RewardFn whose '__call__' method computes the reward of an environment
+                transition. The function must compute the reward based on the current state,
+                the chosen action and the next state.
+                Implement options are ['DenseReward', 'SparseReward']. Defaults to 'DenseReward'.
+            viewer: 'Viewer' used for rendering. Defaults to the 'OPViewer' with 'human' render mode.   
+        """
+        
+        
+        self.generator = generator or generator.UniformGenerator(
+            num_nodes=20,
+            max_length=2,
+        )
+        self.num_nodes = self.generator.num_nodes
+        self.max_length = self.generator.max_length
+        self.reward_fn = reward_fn or reward.DenseReward()
+        self._viewer = viewer #or OPViewer(name="OP", render_mode="human")
+        
+        
+    def __repr__(self) -> str:
+        return f"OP environment with {self.num_nodes} nodes and travel budget of {self.max_length}."
+    
+    
+    def reset(self, key: chex.PRNGKey) -> Tuple[types_.State, TimeStep[types_.Observation]]:
```

Comment:
Same for the types. Please make sure to match the style of the rest of the repository. I'll stop adding comments about this but please make sure to double check all your files.